### PR TITLE
instantsend: various fixes

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

**Comment by UdjinM6** [comment]
ping @PastaPastaPasta

File: src/llmq/instantsend.cpp [hunk 1]
```diff
@@ -345,7 +345,9 @@ uint256 CInstantSendDb::GetInstantSendLockHashByTxid(const uint256& txid) const
     LOCK(cs_db);
     uint256 islockHash;
     if (!txidCache.get(txid, islockHash)) {
-        db->Read(std::make_tuple(DB_HASH_BY_TXID, txid), islockHash);
+        if (!db->Read(std::make_tuple(DB_HASH_BY_TXID, txid), islockHash)) {
+            return {};
```
Comment (PastaPastaPasta) [inline src/llmq/instantsend.cpp:349]: I wonder if it makes more sense to return an std::optional<uint256> as opposed to returning a null / empty value here... I think that would potentially be a less error prone api

File: src/llmq/instantsend.cpp [hunk 2]
```diff
@@ -1051,16 +1062,27 @@ void CInstantSendManager::ProcessInstantSendLock(NodeId from, const uint256& has
...
+        if (sameTxIsLock->IsDeterministic()) {
+            // can happen, nothing to do
+            return;
+        } else if (islock->IsDeterministic()) {
+            // can happen, remove and archive the non-deterministic sameTxIsLock
+            db.RemoveAndArchiveInstantSendLock(sameTxIsLock, WITH_LOCK(::cs_main, return ::ChainActive().Height()));
+        }
+    } else {
+        for (const auto& in : islock->inputs) {
+            auto sameOutpointIsLock = db.GetInstantSendLockByInput(in);
```
Comment (PastaPastaPasta) [inline src/llmq/instantsend.cpp:1081]: nit preferably should be const

File: src/llmq/instantsend.cpp [hunk 3]
```diff
@@ -1051,16 +1062,27 @@ void CInstantSendManager::ProcessInstantSendLock(NodeId from, const uint256& has
...
-    if (otherIsLock != nullptr) {
-        LogPrintf("CInstantSendManager::%s -- txid=%s, islock=%s: duplicate islock, other islock=%s, peer=%d\n", __func__,
-                  islock->txid.ToString(), hash.ToString(), ::SerializeHash(*otherIsLock).ToString(), from);
-    }
-    for (const auto& in : islock->inputs) {
-        otherIsLock = db.GetInstantSendLockByInput(in);
-        if (otherIsLock != nullptr) {
-            LogPrintf("CInstantSendManager::%s -- txid=%s, islock=%s: conflicting input in islock. input=%s, other islock=%s, peer=%d\n", __func__,
-                      islock->txid.ToString(), hash.ToString(), in.ToStringShort(), ::SerializeHash(*otherIsLock).ToString(), from);
+    auto sameTxIsLock = db.GetInstantSendLockByTxid(islock->txid);
```
Comment (PastaPastaPasta) [inline src/llmq/instantsend.cpp:1065]: nit preferably const

File: src/llmq/instantsend.cpp [hunk 4]
```diff
@@ -441,6 +441,17 @@ std::vector<uint256> CInstantSendDb::RemoveChainedInstantSendLocks(const uint256
     return result;
 }
 
+void CInstantSendDb::RemoveAndArchiveInstantSendLock(const CInstantSendLockPtr& islock, int nHeight)
+{
+    LOCK(cs_db);
+
+    CDBBatch batch(*db);
+    auto hash = ::SerializeHash(*islock);
```
Comment (PastaPastaPasta) [inline src/llmq/instantsend.cpp:449]: nit const

**Review by PastaPastaPasta — DISMISSED** [review]
A couple of thoughts, overall though, this PR seems fine

utACK for squash merge

File: src/llmq/instantsend.cpp (hunk 1 above)
Comment (UdjinM6) [reply src/llmq/instantsend.cpp:349]: This feels a bit out of scope imo.

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**